### PR TITLE
Add section on multi-threading to gRPC services doc

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/grpc/grpc-dotnet/issues/1845

[Internal Review](https://review.docs.microsoft.com/en-us/aspnet/core/grpc/services?view=aspnetcore-7.0&branch=pr-en-us-26767)

### CODE REVIEW

**Comment by JamesNK** [comment]
@halter73, @captainsafia, @BrennanConroy, @brunolins16 re: the producer/consumer sample code

File: aspnetcore/grpc/services.md [hunk 1]
```diff
@@ -268,6 +268,68 @@ public override Task<ExampleResponse> UnaryCall(ExampleRequest request, ServerCa
...
 
+## Multi-threading with gRPC streaming methods
+
+There are important considerations to gRPC streaming methods that use multiple threads.
+
+### Reader and writer thread safety
+
+`IAsyncStreamReader<TMessage>` and `IServerStreamWriter<TMessage>` can each be used by only one thread at a time. For a streaming gRPC method, multiple threads can't read new messages with `requestStream.MoveNext()` simultaneously. And multiple threads can't write new messages with `responseStream.WriteAsync(message)` simultaneously.
+
+A safe way to enable multiple threads to interact with a gRPC method is to use the producer/consumer pattern with [System.Threading.Channels](/dotnet/core/extensions/channels).
```
Comment (wadepickett) [inline aspnetcore/grpc/services.md:279]:
```suggestion
A safe way to enable multiple threads to interact with a gRPC method is to use the producer-consumer pattern with [System.Threading.Channels](/dotnet/core/extensions/channels).
```

File: aspnetcore/grpc/services.md [hunk 2]
```diff
@@ -268,6 +268,68 @@ public override Task<ExampleResponse> UnaryCall(ExampleRequest request, ServerCa
...
+* Creates a bounded channel for producing and consuming `DataResult` messages.
+* Writes messages to the channel from multiple background tasks.
+* Reads messages from the channel and writes them to the response stream.
+
+> [!NOTE]
+> Bidirectional streaming methods take `IAsyncStreamReader<TMessage>` and `IServerStreamWriter<TMessage>` as arguments. It's safe to use these types on separate threads from each other.
+
+### Interacting with gRPC method after call ends
+
+A gRPC call ends on the server once the gRPC method exits. Some arguments passed to gRPC methods aren't safe to use after the call has ended:
```
Comment (wadepickett) [inline aspnetcore/grpc/services.md:325]:
```suggestion
A gRPC call ends on the server once the gRPC method exits. The following arguments passed to gRPC methods aren't safe to use after the call has ended:
```

**Review by wadepickett — APPROVED** [review]
Just some really minor nits to consider.  I kept the review narrow, to just the PR changes.

File: aspnetcore/grpc/services.md [hunk 3]
```diff
@@ -268,6 +268,68 @@ public override Task<ExampleResponse> UnaryCall(ExampleRequest request, ServerCa
...
+        channel.Writer.Complete();
+    });
+
+    // Read results from channel.
+    await foreach (var message in channel.Reader.ReadAllAsync())
+    {
+        await responseStream.WriteAsync(message);
+    }
+
+    await processTask;
```
Comment (BrennanConroy) [inline aspnetcore/grpc/services.md:310]: I know it's in a comment below, but maybe a code comment would be good to about making sure to wait for everything to complete before exiting the method.

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**